src/mechanics/camera.rs

```diff
@@ -1,18 +1,489 @@
+
+    const TEST_LEVEL_WIDTH_IN_BOUNDS: f32 = 250.0;
+    const TEST_LEVEL_WIDTH_OUT_LBOUNDS: f32 = -500.0;
+    const TEST_LEVEL_WIDTH_OUT_RBOUNDS: f32 = 1000.0;
+
+    const TEST_LEVEL_HEIGHT_IN_BOUNDS: f32 = 250.0;
+    const TEST_LEVEL_HEIGHT_OUT_TBOUNDS: f32 = 1000.0;
+    const TEST_LEVEL_HEIGHT_OUT_BBOUNDS: f32 = -500.0;
+
+    fn setup_app_bounds_checking() -> App {
+        let mut app = App::new();
+
+        app.insert_resource(LevelDimensions {
+            width: TEST_LEVEL_WIDTH,
+            height: TEST_LEVEL_HEIGHT,
+        });
+
+        app.add_system(move_camera);
+
+        app
+    }
+
+    #[test]
+    fn within_bounds() {
+        let mut app = setup_app_bounds_checking();
+
+        let _player_id = app
+            .world
+            .spawn()
+            .insert_bundle((
+                Player,
+                Transform::from_xyz(
+                    TEST_LEVEL_WIDTH_IN_BOUNDS, 
+                    TEST_LEVEL_HEIGHT_IN_BOUNDS,
+                    0.0),
+            ))
+            .id();
+
+        let camera_id = app
+            .world
+            .spawn()
+            .insert_bundle(Camera2dBundle {
+                projection: OrthographicProjection { 
+                    left: (0.0), 
+                    right: ((CAMERA_WIDTH - 1) as f32), 
+                    bottom: (0.0), 
+                    top: ((CAMERA_HEIGHT - 1) as f32), 
+                    ..default()},
+                ..default()
+            })
+            .id();
+
+        app.update();
+
+        let camera_query = app.world.get::<Transform>(camera_id);
+        assert!(camera_query.is_some());
+
+        let expected_transform_x = TEST_LEVEL_WIDTH_IN_BOUNDS;
+        let expected_transform_y = TEST_LEVEL_HEIGHT_IN_BOUNDS;
+
+        let actual_transform = *camera_query.unwrap();
+        let actual_transform_x = actual_transform.translation.x;
+        let actual_transform_y = actual_transform.translation.y;
+
+        assert_eq!(expected_transform_x, actual_transform_x);
+        assert_eq!(expected_transform_y, actual_transform_y);
+    }
+
+    #[test]
+    fn out_of_bounds_left() {
+        let mut app = setup_app_bounds_checking();
+
+        let _player_id = app
+            .world
+            .spawn()
+            .insert_bundle((
+                Player,
+                Transform::from_xyz(
+                    TEST_LEVEL_WIDTH_OUT_LBOUNDS, 
+                    TEST_LEVEL_HEIGHT_IN_BOUNDS,
+                    0.0),
+            ))
+            .id();
+
+        let camera_id = app
+            .world
+            .spawn()
+            .insert_bundle(Camera2dBundle {
+                projection: OrthographicProjection { 
+                    left: (0.0), 
+                    right: ((CAMERA_WIDTH - 1) as f32), 
+                    bottom: (0.0), 
+                    top: ((CAMERA_HEIGHT - 1) as f32), 
+                    ..default()},
+                ..default()
+            })
+            .id();
```

Comment:
Since this camera spawning code doesn't change per test, what do you think about putting this into its own function?
```rust
fn spawn_camera(app: &mut App) -> Entity {
  let camera_id = app
    .world
    .spawn()
    .insert_bundle(Camera2dBundle {
        projection: OrthographicProjection { 
            left: 0.0, 
            right: (CAMERA_WIDTH - 1) as f32, 
            bottom: 0.0, 
            top: (CAMERA_HEIGHT - 1) as f32, 
            ..default()},
        ..default()
    })
    .id();

    camera_id
}

...
#[test]
fn some_bounds_test() {
    ...

    let camera_id = spawn_camera(&mut app);

    ...
}
```

Reply:
Fantastic optimization